storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
+    }
+
+    buf_block_t *header_block=
+      fsp_get_latched_page(page_id_t{0, 0}, &mtr, &err);
+    if (!header_block)
+      goto err_exit;
+
+    buf_block_t *root= btr_root_block_get(m_index, RW_SX_LATCH, &mtr,
+                                          &err);
+    if (UNIV_UNLIKELY(!root))
+      goto err_exit;
+
+    fseg_header_t *seg_header=
+      root->page.frame + (level
+                          ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+                          : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &mtr,
+                                            &iblock, &err);
+    if (!inode)
+      goto err_exit;
+
+    auto parent_it= m_parent_pages.find(cur_page_no);
+    if (parent_it == m_parent_pages.end())
+    {
+      err= DB_CORRUPTION;
+      goto err_exit;
+    }
+
+    uint32_t parent_page_no= uint32_t(parent_it->second);
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      goto err_exit;
+    }
+
+    pageAllocator allocator(header_block, iblock, inode, extent_size,
+                            &mtr);
+
+    associatedPages related_pages(block, &mtr);
+    err= allocator.prepare(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) goto err_exit;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) goto err_exit;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, &mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr.discard_modifications();
+      goto err_exit;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    mtr.memcpy(*new_block, new_block->page.frame,
+               block->page.frame, srv_page_size - FIL_PAGE_DATA_END);
+    mtr.write<4,mtr_t::FORCED>(*new_block,
+                               FIL_PAGE_OFFSET + new_block->page.frame,
+                               new_page_no);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
```

Comment:
We must use the same `mtr` object for all modifications.

Reply:
we do use same mtr only. Please check